core/chapters/c06_lists.py

```diff
@@ -423,92 +423,319 @@ def check(self):
+            print(len(words))
+
+    class print_last_element(ExerciseStep):
+        """
+Exercise: for any non-empty list `words`, print the last element. For example, if
+
+    __no_auto_translate__
+    words = ['This', 'is', 'a', 'list']
+
+your program should print `list`.
+        """
+
+        hints = """
+To access the last element of the list, you'll need the index of the last position.
+If the list has 2 elements, the first element is at index 0, so the last element is at index 1.
+Likewise, if the list had 3 elements, the last element would be at index 2.
+Do you see a pattern between those numbers? How can you express it?
+Can you come up with a general solution that works for any length?
+        """
+
+        def solution(self, words: List[str]):
+            print(words[len(words) - 1])
+
+        tests = [
+            (["Python"], "Python"),
+            (['Hello', 'world'], "world"),
+            (['futurecoder', 'is', 'cool!'], "cool!"),
+            (['This', 'is', 'a', 'list'], "list")
+        ]
+
+    class range_len(ExerciseStep):
+        """
+So in general, the valid indices are:
 
+    [0, 1, 2, ..., len(words) - 2, len(words) - 1]
+
+Now we can fix the program from earlier to work with any list. Fill in the `...`:
+
+    __copyable__
+    __no_auto_translate__
+    words = ['This', 'is', 'a', 'list']
+
+    for index in ...:
+        print(index)
+        print(words[index])
+
+For the given example value of `words` it should print:
+
+    0
+    This
+    1
+    is
+    2
+    a
+    3
+    list
+        """
+
+        hints = """
+This is similar to an earlier step, where we looped through two lists `words` and `indices` at the same time.
+This time there is only one list, and it could be of any length.
+Think back to earlier steps. We replaced `indices` with a function.
+How can you combine the two functions you just learned, `range` and `len`, to help you here?
+        """
+
+        def solution(self, words: List[str]):
             for index in range(len(words)):
                 print(index)
                 print(words[index])
 
+        tests = [
+                   (['Python'], """\
+0
+Python
+                        """),
+                    (['Hello', 'world'], """\
+0
+Hello
+1
+world
+                    """),
+                    (['futurecoder', 'is', 'cool!'], """\
+0
+futurecoder
+1
+is
+2
+cool!
+                    """),
+                    (['This', 'is', 'a', 'list'], """\
+0
+This
+1
+is
+2
+a
+3
+list
+                    """),
+                ]
+
     class index_exercise(ExerciseStep):
         """
-Let's get some exercise! Given a list `things` and a value `to_find`,
+If you're not quite comfortable with `range` and/or `len`, practice and experiment with it for a bit.
+Here are some simple exercises you can try on your own if you want.
+For any non-empty list `words` and any positive number `n`:
+
+- Print the numbers from `1` to `100` inclusive.
```

Comment:
Hmmm, if you try any of these it shows a message because it's assuming you're doing the actual exercise. It'd be nice if this was the final text and the remaining steps were moved into a new page but that'll mess with existing user data.

Reply:
I think moving them to the final text is a good idea. The challenging exercise will be a bit... challenging this way but it should be fine. `len` is not as confusing as `range`, which we eased them into.

Reply:
what do you think?

Reply:
You'll have to add some code to `book/store.js` to handle old user data in the new page structure. If they're on the step `index_exercise` then change it to `final_text`. If they're on a step after that, change the page slug to the new one.